src/storage/pooled_storage_manager.h

```diff
@@ -89,16 +95,17 @@ void* GPUPooledStorageManager::Alloc(size_t size) {
   }
 }
 
-void GPUPooledStorageManager::Free(void* ptr, size_t size) {
+void GPUPooledStorageManager::Free(void* ptr, size_t raw_size) {
   std::lock_guard<std::mutex> lock(mutex_);
+  size_t size = raw_size + ndev_;
   auto&& reuse_pool = memory_pool_[size];
   reuse_pool.push_back(ptr);
 }
 
 void GPUPooledStorageManager::ReleaseAll() {
   for (auto&& i : memory_pool_) {
     for (auto&& j : i.second) {
-      DirectFree(j, i.first);
+      DirectFree(j, i.first - ndev_);
```

Comment:
This argument is size, not a pointer. In alloc, free and directfree I do size = raw_size + ndev_, so this cancels that (this is because DirectFree is called also from storage.cc, and I did not want to modify that).
The whole thing is done to be able to scatter the allocation into the same number of bytes between gpus - unfortunately I cannot guarantee that you will always use all GPUs, so I can't just do padding to the closest multiply of ndev_. Adding ndev_ makes sure that I can scatter into any number up to ndev_.